daffodil-runtime1/src/main/scala/edu/illinois/ncsa/daffodil/processors/parsers/BCDParsers.scala

```diff
@@ -0,0 +1,103 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH THE
+ * SOFTWARE.
+ */
+
+package edu.illinois.ncsa.daffodil.processors.parsers
+
+import edu.illinois.ncsa.daffodil.processors.Evaluatable
+import edu.illinois.ncsa.daffodil.schema.annotation.props.gen.LengthUnits
+import edu.illinois.ncsa.daffodil.processors.ElementRuntimeData
+import edu.illinois.ncsa.daffodil.util.DecimalUtils.{ bcdToBigInteger, bcdToBigDecimal }
+import java.lang.{ Long => JLong, Number => JNumber }
+import edu.illinois.ncsa.daffodil.processors.ParseOrUnparseState
+
+class BCDDecimalKnownLengthParser(e: ElementRuntimeData, binaryDecimalVirtualPoint: Int, val lengthInBits: Int)
+  extends BCDDecimalParserBase(e, binaryDecimalVirtualPoint)
+  with HasKnownLengthInBits {
+}
+
+class BCDDecimalRuntimeLengthParser(val e: ElementRuntimeData, binaryDecimalVirtualPoint: Int, val lengthEv: Evaluatable[JLong], val lUnits: LengthUnits)
+  extends BCDDecimalParserBase(e, binaryDecimalVirtualPoint)
+  with HasRuntimeExplicitLength {
+}
+
+abstract class BCDDecimalParserBase(e: ElementRuntimeData, binaryDecimalVirtualPoint: Int)
+  extends PrimParserObject(e) {
+
+  protected def getBitLength(s: ParseOrUnparseState): Int
+
+  def parse(start: PState): Unit = {
+    val nBits = getBitLength(start)
+    val dis = start.dataInputStream
+
+    if (!dis.isDefinedForLength(nBits)) {
+      PE(start, "Insufficient bits in data. Needed %d bit(s) but found only %d available.", nBits, dis.remainingBits.get)
+      return
+    }
+
+    val bigDec = bcdToBigDecimal(dis.getByteArray(nBits), binaryDecimalVirtualPoint)
+    start.simpleElement.overwriteDataValue(bigDec)
+  }
+}
+
+class BCDIntegerRuntimeLengthParser(val e: ElementRuntimeData, val lengthEv: Evaluatable[JLong], val lUnits: LengthUnits)
+  extends BCDIntegerBaseParser(e)
+  with HasRuntimeExplicitLength {
+}
+
+class BCDIntegerKnownLengthParser(e: ElementRuntimeData, val lengthInBits: Int)
+  extends BCDIntegerBaseParser(e)
+  with HasKnownLengthInBits {
+}
+
+abstract class BCDIntegerBaseParser(e: ElementRuntimeData)
+  extends PrimParserObject(e) {
+
+  protected def getBitLength(s: ParseOrUnparseState): Int
+
+  def parse(start: PState): Unit = {
+    val nBits = getBitLength(start)
+    if (nBits == 0) return // zero length is used for outputValueCalc often.
+    val dis = start.dataInputStream
+
+    if (!dis.isDefinedForLength(nBits)) {
+      PE(start, "Insufficient bits in data. Needed %d bit(s) but found only %d available.", nBits, dis.remainingBits.get)
+      return
+    }
+
+    // AS400BCDDeimal only has conversion to Double and BigDecimal, need to convert to Number
```

Comment:
Is this comment correct? I don't see any use of AS400BCDecimal? Maybe this is old? Is the nBits > 64 check needed?

Reply:
I just need to remove this comment, it simply got copied around from when I based this code off of packed decimal